File: papras_demo/src/roomba_nav.py
#!/usr/bin/env python3
'''
Roomba Navigation using ARUCO tags.

Twist
    linear:
      x: 0.0
      y: 0.0
      z: 0.0
    angular:
      x: 0.0
      y: 0.0
      z: 0.0
'''

import rospy
from std_msgs.msg import String, Bool
from geometry_msgs.msg import Twist, PoseWithCovariance, PoseStamped
from aruco_msgs.msg import MarkerArray, Marker

# global_aruco_tag_position = [None, None, None]
aruco_tags = None
aruco_tag_ids = None
prev_sequence, curr_sequence = -1, -1
tolerance = 0.025 #m

def get_position_from_marker(marker):
    x,y,z = float(marker.pose.pose.position.x), float(marker.pose.pose.position.y), float(marker.pose.pose.position.z)
    return x,y,z

def aruco_tag_callback(data):
    global aruco_tags, aruco_tag_ids, curr_sequence

    aruco_tags = data.markers
    aruco_tag_ids = []
    for tag in data.markers:
        aruco_tag_ids.append(int(tag.id))
    # x,y,z = float(data.pose.position.x), float(data.pose.position.y), float(data.pose.position.z)
    # global_aruco_tag_position = [x,y,z]

    curr_sequence = int(data.header.seq)

def find_tag(publisher, goal_tag_id):
    global aruco_tag_ids, prev_sequence, curr_sequence
    vel_cmd = Twist()

    while prev_sequence == curr_sequence or goal_tag_id not in aruco_tag_ids:
        # no updated tag position is being received (i.e. the camera does not detect the aruco tag)
        # or can't find specific aruco tag id
        vel_cmd.angular.z = 0.15
        publisher.publish(vel_cmd)

    vel_cmd.angular.z = 0.0
    publisher.publish(vel_cmd)

    prev_sequence = curr_sequence

def roomba_nav(publisher, goal_tag_id):
    global aruco_tags, aruco_tag_ids, tolerance

    if goal_tag_id not in aruco_tag_ids:
        print("Can't find the goal aruco tag. Leave nav to stall in search.")

    x,y,z = None, None, None
    for tag in aruco_tags:
        if int(tag.id) == goal_tag_id:
            x,y,z = get_position_from_marker(tag)
            break

    if x == None:
        print("Error! Something is wrong. Leave nav to stall in search.")

    vel_cmd = Twist()
    vel_cmd.linear.x = 0
    vel_cmd.angular.z = 0

    if x > tolerance:
        # aruco tag to the right of the camera
        vel_cmd.angular.z = -0.1
    elif x < -tolerance:
        # aruco tag to the left of the camera
        vel_cmd.angular.z = 0.1

    if z > 0.2:
        # aruco tag is more than 2m away
        vel_cmd.linear.x = 0.2
    elif z > 0.12:
        # aruco tag is more than 1m away
        vel_cmd.linear.x = 0.1
    elif z < 0.08:
        # aruco tag is too close, less than 0.5m away
        vel_cmd.linear.x = -0.1

    publisher.publish(vel_cmd)

def main():
    publisher = rospy.Publisher('/roomba/cmd_vel', Twist, queue_size=10)
    subscriber = rospy.Subscriber("/aruco_tracker/markers", MarkerArray, aruco_tag_callback)
    rospy.init_node('roomba_nav', anonymous=True)
    rate = rospy.Rate(10) # 10hz

    goal_tag_id = 0

    while not rospy.is_shutdown():

        find_tag(publisher, goal_tag_id)
        roomba_nav(publisher, goal_tag_id)

        rate.sleep()

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        print("Error: Failed to start ROOMBA Navigation!")

File: papras_demo/src/roomba_nav.cpp
/* Authors: Kazuki Shin, Sankalp Yamsani */

// ROS
#include <ros/ros.h>

// TF2
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
#include <tf/transform_listener.h>
#include <tf2_ros/transform_broadcaster.h>
#include <aruco_msgs/MarkerArray.h>
#include <aruco_msgs/Marker.h>

// eigen_conversions
#include <eigen_conversions/eigen_msg.h>

// MoveIt!
#include <geometric_shapes/solid_primitive_dims.h>
#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit/planning_scene_interface/planning_scene_interface.h>
#include <moveit/planning_scene_monitor/planning_scene_monitor.h>
#include <moveit_visual_tools/moveit_visual_tools.h>


// Other
#include <sstream>
#include <std_msgs/String.h>
#include <std_srvs/Empty.h>
#include <std_msgs/Float32.h>
#include <std_msgs/Bool.h>
#include <rosparam_shortcuts/rosparam_shortcuts.h>
#include <geometry_msgs/Twist.h>
#include <geometry_msgs/Pose.h>
#include <cmath>        // std::abs
#include <vector>
#include <unordered_map>




enum state
{
    ST_INIT,
    ST_PAUSED,
    ST_ARM_TO_REST_START,
    ST_INIT_SIM_OBJ,
    ST_CAPTURE_OBJ,
    ST_PICK_OBJ,
    ST_ARM_TO_TRANSPORT,
    ST_PLACE_OBJ,
    ST_ARM_TO_HOME_END,
    ST_DONE
};

int goal_tag_id = 0;
aruco_msgs::MarkerArray found_markers;
std::unordered_map<int,int> marker_id_index;
int camera_upside_down = -1;
double tolerance = 0.22;

int aruco_tag_cb_cnt = -1;
bool finished_roomba_nav = false;
volatile bool start_roomba = false;

geometry_msgs::Pose get_position_from_marker(aruco_msgs::Marker marker){
  return  marker.pose.pose;
}

void aruco_tag_callback(const aruco_msgs::MarkerArray::ConstPtr& msgs){
  ROS_INFO_STREAM(*msgs);
  found_markers = *msgs;
  marker_id_index.erase(marker_id_index.begin(),marker_id_index.end());
  for(int i = 0; i < found_markers.markers.size(); i++){
    marker_id_index[(int)found_markers.markers[i].id] = i;
  }
  if (marker_id_index.find(goal_tag_id) != marker_id_index.end())
    aruco_tag_cb_cnt = 0;
}

void start_roomba_callback(const std_msgs::Bool data){
    start_roomba = data.data;
}

void find_tag(ros::Publisher pub,ros::NodeHandle nh ){
  geometry_msgs::Twist cmd_vel_msg;
  cmd_vel_msg.angular.z = 0.4;
  pub.publish(cmd_vel_msg);
}

void roomba_nav(ros::Publisher pub, int goal){
  ROS_INFO_STREAM("in roomba nav goal id: " << goal);
  if(marker_id_index.find(goal) == marker_id_index.end()){
    return;
  }
  geometry_msgs::Twist cmd_vel_msg;
  geometry_msgs::Pose p =  get_position_from_marker(found_markers.markers[marker_id_index[goal]]);

  if(p.position.x == NULL){
    return;
  }

  cmd_vel_msg.linear.x = 0.0;
  cmd_vel_msg.angular.z = 0.0;

  if(p.position.x > tolerance){
    cmd_vel_msg.angular.z = (-0.2 * camera_upside_down) * (1+(std::abs(p.position.x-tolerance)));
  }else if (p.position.x < - tolerance){
    cmd_vel_msg.angular.z = (0.2 * camera_upside_down) * (1+(std::abs(p.position.x-tolerance)));
  }

  if (goal_tag_id == 0)
  {
    if(p.position.z  > 2.0){
      cmd_vel_msg.linear.x = 0.05 * (1+(std::abs(p.position.z-2.0)));
    }else if(p.position.z > 1.25){
      cmd_vel_msg.linear.x = 0.01 * (1+(std::abs(p.position.z - 1.25)));
      tolerance = 0.025;
    }else if(p.position.z  < 1.0){
      cmd_vel_msg.linear.x = -0.01 * (1+(std::abs(p.position.z - 1.00)));
    }else{
      goal_tag_id = 10;
    }
  } else if (goal_tag_id == 10) {
    if(p.position.z  > 2.0){
    cmd_vel_msg.linear.x = 0.05 * (1+(std::abs(p.position.z-2)));
    }else if(p.position.z > 0.63){
      cmd_vel_msg.linear.x = 0.05 * (1+(std::abs(p.position.z - 0.63)));
      tolerance = 0.15;
    }else if(p.position.z  < 0.62){
      cmd_vel_msg.linear.x = -0.05 * (1+(std::abs(p.position.z -0.62)));
    }else{
      finished_roomba_nav = true;      
    }
  }

  // pub.publish(cmd_vel_msg);
}

int main(int argc, char** argv)
{
  ros::init(argc, argv, "roomba_nav");
  ros::AsyncSpinner spinner(2);
  spinner.start();
  ros::Rate r(10); 
  ros::NodeHandle nh;
  state task_state = ST_INIT;
  state paused_state = ST_INIT;

  std::string world_name;
  bool handover_planned;

  // Load rosparams
  ros::NodeHandle nh_priv("~");
  std::string param_path;

  moveit::planning_interface::MoveGroupInterface group("arm1");
  group.setPlanningTime(45.0);
  group.setPlannerId("RRTConnect");
  group.setMaxAccelerationScalingFactor(0.10);
  group.setMaxVelocityScalingFactor(0.10);

  ros::Publisher cmd_vel_pub = nh.advertise<geometry_msgs::Twist>("/roomba/cmd_vel", 10);
  ros::Publisher start_kitchen = nh.advertise<std_msgs::Bool>("/roomba/start_kitchen", 10);

  ros::Subscriber marker_sub = nh.subscribe("/aruco_marker_publisher/markers", 10, aruco_tag_callback);
  ros::Subscriber handoff_done = nh.subscribe("/big_table/start_roomba",10, start_roomba_callback);
  moveit::planning_interface::MoveGroupInterface::Plan plan;

  ROS_INFO_STREAM("after init");
  goal_tag_id = 10;

  group.setStartStateToCurrentState();
  group.setNamedTarget("plate_big_table");
  moveit::core::MoveItErrorCode error_code = group.plan(plan);
  error_code = group.execute(plan);

  while(!start_roomba);

  group.setStartStateToCurrentState();
  group.setNamedTarget("plate_down");
  error_code = group.plan(plan);
  error_code = group.execute(plan);

  // geometry_msgs::Twist cmd_vel_msg;
  // for(int i = 0; i < 100 && aruco_tag_cb_cnt == -1; i++){
  //   cmd_vel_msg.linear.x = 0.40;
  //   cmd_vel_msg.angular.z = -0.08;
  //   cmd_vel_pub.publish(cmd_vel_msg);
  //   r.sleep();
  // }

  while(ros::ok() && !finished_roomba_nav){
    aruco_tag_cb_cnt+=1;
    // if (aruco_tag_cb_cnt > 10) 
    // {
    //   find_tag(cmd_vel_pub, nh);
    //   ros::Duration(1.0).sleep();
    // }
    roomba_nav(cmd_vel_pub,goal_tag_id);
    r.sleep();
  }

  group.setStartStateToCurrentState();
  group.setNamedTarget("plate_big_table");
  error_code = group.plan(plan);
  error_code = group.execute(plan);

  std_msgs::Bool finished_nav;
  finished_nav.data = true;
  start_kitchen.publish(finished_nav);

  return 0;
}